Add tests to language part 3

bingbing8 opened the pull request:
Add tests for deserialized types and methods, dynamic parameters, generic types, Lacline ending, myinvocation, ordered attribute of hash table. Please review. Thanks!

msftclas commented:
Hi __@bingbing8__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <p>
        It looks like you're a Microsoft contributor (Yanbing Wang). If you're full-time, we DON'T require a Contribution License Agreement. If you are a vendor, please DO sign the electronic Contribution License Agreement. It will take 2 minutes and there's no faxing! https://cla.microsoft.com.
    </p>

TTYL, MSBOT;

JamesWTruher commented on test/powershell/Language/Scripting/DeserializedMethods.Tests.ps1:
```diff
@@ -0,0 +1,40 @@
+
+    AfterAll {
+	        Remove-Item $temp -force -ea silentlycontinue
+        }    
```
please use TESTDRIVE: which is cleaned up automatically

bingbing8 replied:
Fixed.

---
In reply to: [76680218](https://github.com/Powershell/Powershell/pull/2118#discussion_r76680218) [](ancestors = 76680218)

JamesWTruher commented on test/powershell/Language/Scripting/DeserializedTypeConversion.Tests.ps1:
```diff
@@ -0,0 +1,271 @@
+
+    BeforeAll {
+    # Test Import/Export Clixml scenario.
+    $tempFile = [System.IO.Path]::GetTempFileName()
```
please use $TESTDRIVE or TESTDRIVE:

bingbing8 replied:
fixed

---
In reply to: [76680334](https://github.com/Powershell/Powershell/pull/2118#discussion_r76680334) [](ancestors = 76680334)

JamesWTruher commented on test/powershell/Language/Scripting/DeserializedTypeConversion.Tests.ps1:
```diff
@@ -0,0 +1,271 @@
+    AfterAll {        
+        Remove-Item $tempFile -force -ea silentlycontinue
+        # Clean up.
+        gsn | rsn
```
> gsn | rsn [](start = 7, length = 10)

no aliases, please

bingbing8 replied:
fixed.

---
In reply to: [76680408](https://github.com/Powershell/Powershell/pull/2118#discussion_r76680408) [](ancestors = 76680408)

JamesWTruher commented on test/powershell/Language/Scripting/Dynamicparameters.Tests.ps1:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    It "No dynamic parameter shouldn't cause an errr " {        
+        (foo-bar -path def ) | Should Be 'no dynamic parameters'
```
>( [](start = 8, length = 1)

i don't believe the parens are needed

bingbing8 replied:
fixed

---
In reply to: [76680670](https://github.com/Powershell/Powershell/pull/2118#discussion_r76680670) [](ancestors = 76680670)

JamesWTruher commented on test/powershell/Language/Scripting/Generics.Tests.ps1:
```diff
@@ -0,0 +1,74 @@
+
+    It 'dictionaryType[dictionary[list[int],string], stack[double]]' {
+        $x = new-object "$dictionaryType[$dictionaryType[$listType[int],string], $stackType[double]]"
+        $x.gettype().fullname -like '*double*' | Should Be $true
```
>.gettype().fullname -like '*double*' | Should Be $true [](start = 10, length = 54)

`$x.gettype().fullname | should match "double"`

bingbing8 replied:
fixed

---
In reply to: [76681174](https://github.com/Powershell/Powershell/pull/2118#discussion_r76681174) [](ancestors = 76681174)

JamesWTruher commented on test/powershell/Language/Scripting/Generics.Tests.ps1:
```diff
@@ -0,0 +1,74 @@
+            Throw "Execution OK"
+        } catch {
+            $_.FullyQualifiedErrorId | Should be 'TypeNotFoundWithMessage'
+            $_ -like "*`[T`]*" | Should Be $true
```
>$_ -like "*`[T`]*" | Should Be $true [](start = 12, length = 36)

`$_ | should match "\[T\]"`

bingbing8 replied:
fixed

---
In reply to: [76681265](https://github.com/Powershell/Powershell/pull/2118#discussion_r76681265) [](ancestors = 76681265)

JamesWTruher commented:
>         ($strBytes | ? { $_ -eq $crByte } | measure-object).Count | Should Be 5

couldn't this just be `@($strBytes | ?{ $_ -eq $crByte}).Count | should be 5` - why use measure object?

---
Refers to: test/powershell/Language/Scripting/MacLineEndings.Tests.ps1:62 in 9ef2ca0. [](commit_id = 9ef2ca0d45f488d55e115fc7c894f0a70cff321d, deletion_comment = False)

JamesWTruher commented on test/powershell/Language/Scripting/MyInvocation.Tests.ps1:
```diff
@@ -0,0 +1,49 @@
+
+        $a = . {$MyInvocation.Line}
+
+        It 'Script block1' { $a.IndexOf('$a = . {$MyInvocation.Line}') | Should BeGreaterThan -1 }
```
I would be happier if these were:
```
It 'Script block1' {
$a = .  { $MyInvocation.Line }
$a.IndexOf('$a = . {$MyInvocation.Line}' | should begreaterthan -1
}
```

bingbing8 replied:
fixed

---
In reply to: [76685111](https://github.com/Powershell/Powershell/pull/2118#discussion_r76685111) [](ancestors = 76685111)

JamesWTruher commented:
:clock1:

bingbing8 commented:
>         ($strBytes | ? { $_ -eq $crByte } | measure-object).Count | Should Be 5

fixed

---
In reply to: [243247072](https://github.com/Powershell/Powershell/pull/2118#issuecomment-243247072) [](ancestors = 243247072)

---
Refers to: test/powershell/Language/Scripting/MacLineEndings.Tests.ps1:62 in 9ef2ca0. [](commit_id = 9ef2ca0d45f488d55e115fc7c894f0a70cff321d, deletion_comment = False)

lzybkr commented on test/powershell/Language/Scripting/DeserializedMethods.Tests.ps1:
```diff
@@ -0,0 +1,40 @@
+	    $null = $x.Add(456)
+
+	    export-clixml -in $x -path $file
+	    $d = import-clixml $file	
```
You can avoid a file and use the serialization api:

```powershell
$s = [System.Management.Automation.PSSerializer]::Serialize($x)
$d = [System.Management.Automation.PSSerializer]::Deserialize($s)
```

bingbing8 replied:
fixed in Iteration 7.

---
In reply to: [76884878](https://github.com/Powershell/Powershell/pull/2118#discussion_r76884878) [](ancestors = 76884878)

lzybkr commented on test/powershell/Language/Scripting/DeserializedMethods.Tests.ps1:
```diff
@@ -0,0 +1,40 @@
+	    {
+		    $null = $d.Add(789)
+            Throw "Execution OK"
+	    }
```
We should be using better messages, it is confusing when the error logs say "ExecutionOK".

A better error message might be "exception expected, execution should not have reached here"

bingbing8 replied:
fixed.

---
In reply to: [76885029](https://github.com/Powershell/Powershell/pull/2118#discussion_r76885029) [](ancestors = 76885029)

lzybkr commented on test/powershell/Language/Scripting/DeserializedTypeConversion.Tests.ps1:
```diff
@@ -0,0 +1,271 @@
+    } 
+
+    It 'T1' { 
+
```
We should use a more descriptive test name.

bingbing8 replied:
fixed

---
In reply to: [76885462](https://github.com/Powershell/Powershell/pull/2118#discussion_r76885462) [](ancestors = 76885462)

lzybkr commented on test/powershell/Language/Scripting/DeserializedTypeConversion.Tests.ps1:
```diff
@@ -0,0 +1,271 @@
+
+    It 'T1 Test remoting scenario.'  -skip:$IsCoreCLR {
+        $s = New-PSSession
+
```
Are we adding useful coverage here? Remoting can be slow and isn't always reliable. I'd like to keep remoting out of tests where possible.

bingbing8 replied:
removed remote cases. we are testing the deserialized type. it does not matter how we get the type.

---
In reply to: [76885902](https://github.com/Powershell/Powershell/pull/2118#discussion_r76885902) [](ancestors = 76885902)

lzybkr commented on test/powershell/Language/Scripting/DeserializedTypeConversion.Tests.ps1:
```diff
@@ -0,0 +1,271 @@
+﻿Describe "Tests conversion of deserialized types to original type using object properties." -Tags "CI" {
+
```
It would be good to add coverage for PowerShell classes as well.

bingbing8 replied:
Added two test cases for powershell classes


---
In reply to: [76886085](https://github.com/Powershell/Powershell/pull/2118#discussion_r76886085) [](ancestors = 76886085)

lzybkr commented on test/powershell/Language/Scripting/Dynamicparameters.Tests.ps1:
```diff
@@ -0,0 +1,68 @@
+﻿Describe "Dynamic parameter support in script cmdlets." -Tags "CI" {
+    BeforeAll {
```
We should also add some tests where the script cmdlet returns a bindable object, basically a class that has properties with the Parameter attribute.

bingbing8 replied:
added

---
In reply to: [76886461](https://github.com/Powershell/Powershell/pull/2118#discussion_r76886461) [](ancestors = 76886461)

lzybkr commented on test/powershell/Language/Scripting/Generics.Tests.ps1:
```diff
@@ -0,0 +1,74 @@
+﻿Describe "Generics support" -Tags "CI" {
+    #
```
we should add coverage for use of PowerShell classes as type arguments.

bingbing8 replied:
added

---
In reply to: [76886683](https://github.com/Powershell/Powershell/pull/2118#discussion_r76886683) [](ancestors = 76886683)

daxian-dbw commented on test/powershell/Language/Scripting/DeserializedTypeConversion.Tests.ps1:
```diff
@@ -0,0 +1,284 @@
+﻿Describe "Tests conversion of deserialized types to original type using object properties." -Tags "CI" {
+
+    BeforeAll {    
```
Can you please fix the indentation in this `BeforeAll` section?

bingbing8 replied:
fixed

---
In reply to: [77376739](https://github.com/Powershell/powershell/pull/2118#discussion_r77376739) [](ancestors = 77376739)

daxian-dbw commented on test/powershell/Language/Scripting/Dynamicparameters.Tests.ps1:
```diff
@@ -0,0 +1,85 @@
+                $paramDictionary.name
+            }
+            elseif ($paramDictionary -ne $null) {
+              if ($paramDictionary.dp1.Value -ne $null) {
```
And indentation in this `elseif` block.

bingbing8 replied:
fixed

---
In reply to: [77376868](https://github.com/Powershell/powershell/pull/2118#discussion_r77376868) [](ancestors = 77376868)

daxian-dbw commented on test/powershell/Language/Scripting/OrderedAttributeForHashTables.Tests.ps1:
```diff
@@ -0,0 +1,70 @@
+	   $a = new-object psobject -property ([ordered]@{one=1;two=2})
+
+       $a | Should Not Be $null
+       $a.one | Should Be 1
```
And indentation in this `It` block.

bingbing8 replied:
fixed al indentation in this request

---
In reply to: [77377165](https://github.com/Powershell/powershell/pull/2118#discussion_r77377165) [](ancestors = 77377165)

daxian-dbw commented on test/powershell/Language/Scripting/OrderedAttributeForHashTables.Tests.ps1:
```diff
@@ -0,0 +1,70 @@
+</helpItems>
+'@
+
+            { $script:a = select-xml -content $helpXml -xpath "//command:name" -namespace (
```
And indentation in this `Context` block.

bingbing8 replied:
fixed

---
In reply to: [77377223](https://github.com/Powershell/powershell/pull/2118#discussion_r77377223) [](ancestors = 77377223)

daxian-dbw commented on test/powershell/Language/Scripting/OrderedAttributeForHashTables.Tests.ps1:
```diff
@@ -0,0 +1,70 @@
+
+           $script:a = $null
+           
+	       {$script:a = dir | select-object -property Name, ([ordered]@{Name="IsDirectory";Expression ={$_.PSIsContainer}})} | Should Not Throw        
```
ditto.

bingbing8 replied:
fixed

---
In reply to: [77377282](https://github.com/Powershell/powershell/pull/2118#discussion_r77377282) [](ancestors = 77377282)

daxian-dbw commented on test/powershell/Language/Scripting/DeserializedMethods.Tests.ps1:
```diff
@@ -0,0 +1,27 @@
+	    $null = $x.Add(456)
+
+        $s = [System.Management.Automation.PSSerializer]::Serialize($x)
+        $d = [System.Management.Automation.PSSerializer]::Deserialize($s)	    
```
can you please fix the indentation here?

bingbing8 replied:
this one already fixed. IE did not show the updated code.

---
In reply to: [77377338](https://github.com/Powershell/powershell/pull/2118#discussion_r77377338) [](ancestors = 77377338)

bingbing8 replied:
fixed.

---
In reply to: [78826306](https://github.com/Powershell/Powershell/pull/2118#discussion_r78826306) [](ancestors = 78826306)

daxian-dbw commented:
@bingbing8 could you please fix the indentation in the changes? This PR can be merged as soon as you fix the indentation problem.

bingbing8 commented on test/powershell/Language/Scripting/OrderedAttributeForHashTables.Tests.ps1:
```diff
@@ -0,0 +1,70 @@
+       
+           $script:a = $null       
+           
+	         { $script:a = set-wmiinstance -class win32_environment -argument ([ordered]@{Name="TestWmiInstance";
```
fixed

---
In reply to: [77377264](https://github.com/Powershell/powershell/pull/2118#discussion_r77377264) [](ancestors = 77377264)

bingbing8 commented:
fixed.

---
In reply to: [244429037](https://github.com/Powershell/powershell/pull/2118#issuecomment-244429037) [](ancestors = 244429037)

lzybkr commented on test/powershell/Language/Scripting/DeserializedMethods.Tests.ps1:
```diff
@@ -0,0 +1,27 @@
+﻿Describe "DeserializedMethods" -Tags "CI" {    
+    It "Deserialized objects shouldn't ever have any methods (unless they are primitive known types)" {
+        $a = new-object collections.arraylist
```
I think we should use the `[type]::new()` syntax in tests other than testing `New-Object` because it is much faster.

bingbing8 replied:
fixed

---
In reply to: [78832342](https://github.com/Powershell/Powershell/pull/2118#discussion_r78832342) [](ancestors = 78832342)

lzybkr commented on test/powershell/Language/Scripting/DeserializedMethods.Tests.ps1:
```diff
@@ -0,0 +1,27 @@
+	    
+	    try
+	    {
+		    $null = $d.Add(789)
```
We could test this without exceptions by using `Get-Member -MemberType *Method* Add | Should Be $null` or something like that.

bingbing8 replied:
good idea. fixed.

---
In reply to: [78833833](https://github.com/Powershell/Powershell/pull/2118#discussion_r78833833) [](ancestors = 78833833)

lzybkr commented on test/powershell/Language/Scripting/DeserializedTypeConversion.Tests.ps1:
```diff
@@ -0,0 +1,283 @@
+﻿Describe "Tests conversion of deserialized types to original type using object properties." -Tags "CI" {
+    BeforeAll {    
+        # Create new types and test functions.
+        $type1 = Add-Type -PassThru -TypeDefinition @'
```
I think this could be one call to Add-Type like this:

```powershell
$type1,$type2,$type3,$type4 = Add-Type -PassThru -TypeDefinition @'
...
'@
```

bingbing8 replied:
fixed.

---
In reply to: [78838200](https://github.com/Powershell/Powershell/pull/2118#discussion_r78838200) [](ancestors = 78838200)

lzybkr commented on test/powershell/Language/Scripting/DeserializedTypeConversion.Tests.ps1:
```diff
@@ -0,0 +1,283 @@
+            }
+        }
+
+        function test-1
```
This function is called once so it could be moved closer to where it is called.

bingbing8 replied:
fixed

---
In reply to: [78838599](https://github.com/Powershell/Powershell/pull/2118#discussion_r78838599) [](ancestors = 78838599)

lzybkr commented on test/powershell/Language/Scripting/DeserializedTypeConversion.Tests.ps1:
```diff
@@ -0,0 +1,283 @@
+            $test | Format-List | Out-String
+        }
+
+        $t1 = new-object test1 -Property @{name="TestName1";port=80;scriptText="1..5"}
```
The variable is only used once, so it should be moved closer to the use.

bingbing8 replied:
fixed

---
In reply to: [78838934](https://github.com/Powershell/Powershell/pull/2118#discussion_r78838934) [](ancestors = 78838934)

lzybkr commented on test/powershell/Language/Scripting/DeserializedTypeConversion.Tests.ps1:
```diff
@@ -0,0 +1,283 @@
+        }
+    }
+
+    Context 'Type conversion and parameter binding of deserilized Powershell class with default constructor' {
```
s/deserilized/deserialized/

bingbing8 replied:
fixed the this typo in the whole file

---
In reply to: [78839104](https://github.com/Powershell/Powershell/pull/2118#discussion_r78839104) [](ancestors = 78839104)

alexandair replied:
Powershell > PowerShell

lzybkr commented on test/powershell/Language/Scripting/Generics.Tests.ps1:
```diff
@@ -0,0 +1,95 @@
+    It 'list[Int]' {
+        $x = [system.collections.generic.list[int]]::New()
+        $x.Add(42)
+        $x.count | Should Be 1
```
We should test more than 1 item in the collection because non-collections treat count specially, e.g.
```powershell
PS> (42).Count -eq 1
True
```

bingbing8 replied:
fixed

---
In reply to: [78839631](https://github.com/Powershell/Powershell/pull/2118#discussion_r78839631) [](ancestors = 78839631)

lzybkr commented on test/powershell/Language/Scripting/Generics.Tests.ps1:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    It 'dictionaryType[dictionary[list[int],string], stack[double]]' {
+        $x = new-object "$dictionaryType[$dictionaryType[$listType[int],string], $stackType[double]]"
```
With `using namespace`, you could avoid the use of strings in the types here and keep the code clean, assuming you also switch to `::new()`.

lzybkr replied:
On the other hand, it's good to have a generics test with New-Object, so maybe it's fine as is.

bingbing8 replied:
fixed

---
In reply to: [78840129](https://github.com/Powershell/Powershell/pull/2118#discussion_r78840129) [](ancestors = 78840129)

bingbing8 replied:
I tested both.

---
In reply to: [78840273](https://github.com/Powershell/Powershell/pull/2118#discussion_r78840273) [](ancestors = 78840273)

lzybkr commented on test/powershell/Language/Scripting/Generics.Tests.ps1:
```diff
@@ -0,0 +1,95 @@
+        $x | Should match "static\s+TOutput\[\]\s+ConvertAll\[TInput,\s+TOutput\]\("
+    }
+
+   It 'Class' {
```
Can you be more descriptive - `Class` is too generic.

bingbing8 replied:
fixed.

---
In reply to: [78840821](https://github.com/Powershell/Powershell/pull/2118#discussion_r78840821) [](ancestors = 78840821)

lzybkr commented on test/powershell/Language/Scripting/Generics.Tests.ps1:
```diff
@@ -0,0 +1,95 @@
+        }
+    }
+
+    It 'Array' -skip:$IsCoreCLR{
```
Can you be more specific? `Array` is too generic.

bingbing8 replied:
fixed.

---
In reply to: [78840864](https://github.com/Powershell/Powershell/pull/2118#discussion_r78840864) [](ancestors = 78840864)

lzybkr commented on test/powershell/Language/Scripting/MacLineEndings.Tests.ps1:
```diff
@@ -0,0 +1,69 @@
+﻿Describe 'Tests for mac line endings' -Tags "CI" {
+    BeforeAll {        
+        $newline = "`n"
```
Instead of using the same line ending for all tests, I'd put the line ending in `$testData` and use:

CR
LF
CR + LF

CR isn't too interesting (Mac OS up to 9), but the parser handles it fine, so it's good to test.  And maybe rename the file/`Describe` to just `line endings`.

bingbing8 replied:
fixed

---
In reply to: [78842007](https://github.com/Powershell/Powershell/pull/2118#discussion_r78842007) [](ancestors = 78842007)

lzybkr commented on test/powershell/Language/Scripting/MyInvocation.Tests.ps1:
```diff
@@ -0,0 +1,50 @@
+﻿Describe 'Testing of MyInvocation' -Tags "CI" {
+    It 'Script file' {
+
+        $MyInvocation.Line.ToLower().IndexOf('$null = & $scriptblock @parameters') | Should BeGreaterThan -1
```
This test is sensitive to code in Pester because you're matching a line of code from Pester. This is fragile, you should change it to match something in this file, e.g.
```powershell
(<#FOO#> & { $MyInvocation.Line }) | Should Match 'FOO'
```

lzybkr replied:
Maybe this test isn't needed, it's really covered by lines 32 and 36.

bingbing8 replied:
fixed

---
In reply to: [78846919](https://github.com/Powershell/Powershell/pull/2118#discussion_r78846919) [](ancestors = 78846919)

bingbing8 replied:
removed this test.

---
In reply to: [78846400](https://github.com/Powershell/Powershell/pull/2118#discussion_r78846400) [](ancestors = 78846400)

lzybkr commented on test/powershell/Language/Scripting/MyInvocation.Tests.ps1:
```diff
@@ -0,0 +1,50 @@
+
+    Context 'Script block' {        
+
+        It 'Script block1' {
```
These `It` descriptions should be more descriptive.

bingbing8 replied:
fixed

---
In reply to: [78847154](https://github.com/Powershell/Powershell/pull/2118#discussion_r78847154) [](ancestors = 78847154)

bingbing8 committed:
Add tests for the Language

DeserializedMethods, DeserializedTypeConversion,
DynamicParameters, Generics, Indexer, LindEndings, MyInvocation,
OrderedAttributeForHashTable.
